File: day7/haskell/part1.hs
import System.Environment (getArgs)
import qualified Data.Set as Set

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "input.txt" else head args
    content <- readFile filename
    let grid = lines content
        cols = length (head grid)
        startCol = head [c | c <- [0..cols-1], grid !! 0 !! c == 'S']
        result = simulate grid (Set.singleton (0, startCol)) Set.empty
    print (Set.size result)

simulate :: [String] -> Set.Set (Int, Int) -> Set.Set (Int, Int) -> Set.Set (Int, Int)
simulate grid beams hitSplitters
    | Set.null beams = hitSplitters
    | otherwise = simulate grid newBeams newHitSplitters
  where
    results = map (moveDown grid hitSplitters) (Set.toList beams)
    newHits = Set.fromList [pos | (Just pos, _) <- results]
    newHitSplitters = Set.union hitSplitters newHits
    newBeams = Set.fromList (concatMap snd results)

moveDown :: [String] -> Set.Set (Int, Int) -> (Int, Int) -> (Maybe (Int, Int), [(Int, Int)])
moveDown grid hitSplitters (r, c) = go (r + 1)
  where
    rows = length grid
    cols = length (head grid)

    go row
        | row >= rows = (Nothing, [])
        | grid !! row !! c == '^' =
            if Set.member (row, c) hitSplitters
            then (Nothing, [])
            else (Just (row, c), leftRight)
        | otherwise = go (row + 1)
      where
        leftRight = [(row, c - 1) | c > 0] ++ [(row, c + 1) | c < cols - 1]

File: day7/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, _ := os.Open(filename)
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	rows := len(grid)
	cols := len(grid[0])

	// Find S
	startCol := 0
	for c := 0; c < cols; c++ {
		if grid[0][c] == 'S' {
			startCol = c
			break
		}
	}

	// Active beams: use map as set to dedupe
	type pos struct{ r, c int }
	beams := map[pos]bool{{0, startCol}: true}
	hitSplitters := map[pos]bool{}

	for len(beams) > 0 {
		newBeams := map[pos]bool{}

		for beam := range beams {
			r, c := beam.r, beam.c

			// Move beam down
			for r = r + 1; r < rows; r++ {
				if grid[r][c] == '^' {
					splitterPos := pos{r, c}
					if !hitSplitters[splitterPos] {
						hitSplitters[splitterPos] = true
						if c > 0 {
							newBeams[pos{r, c - 1}] = true
						}
						if c < cols-1 {
							newBeams[pos{r, c + 1}] = true
						}
					}
					break
				}
			}
		}

		beams = newBeams
	}

	fmt.Println(len(hitSplitters))
}
